src/documentation/0000-socket-programming-with-net-module/index.md

```diff
@@ -0,0 +1,136 @@
+
+> TCP is a reliable, ordered, and error-checked delivery of a stream of bytes between applications communicating via an internet protocol (IP) network.
+>
+> ~ Wikipedia
+
+## Net Module - Getting Started
+
+The key part of the `net` module is the `net.Socket` class. A socket instance in Node.js is both a duplex stream and an event emitter. This means it is readable, writable, and can utilize the event loop. Sockets can be created by the user and by Node.js, and they are the programmatic object for communicating. Sockets must connect to a `net.Server` class instance; a socket connection to a server is often referred to as a _client connection_. Every socket connection is considered "private" in the sense that the only entities with access to the connection is the server and the client. A server instance **can** have multiple client connections, but will maintain separate communication streams with each connection.
+
+Lets look at the simplest example of a TCP server, an echo server. An _echo server_ is one which returns the same message sent to it.
+
+```js
+const net = require('net')
+
+// Create a TCP server using the `net.createServer()` method
+// The callback method is an event listener for the 'connection' event
+// This method is passed a single parameter, the client socket connection
+const server = net.createServer(client => {
+
+  console.log('Client connected!')
+
+  // The client is a stream, set the data encoding to utf8
+  client.setEncoding('utf8')
+
+  // Manual implementation of an echo server:
+  client.on('data', data => { // data is a UTF-8 string
+    console.log(data) // log the message from the client to the server
+    client.write(data) // write the same message back to the client
+  })
+
+  // Automatic implementation of an echo server:
+  client.pipe(client) // client is a stream, use the Stream.pipe prototype method
+})
+
+// Start the server on localhost:8124
+// The `http` module, inherits this method for its own http.Server.listen
+server.listen(8124, 'localhost', () => {
```

Comment:
```suggestion
server.listen(8124, '127.0.0.1', () => {
```

is `127.0.0.1` more reliable across different systems?

It looks like in other example we've used that rather than `localhost`